fgpyo/fasta/sequence_dictionary.py

```diff
@@ -0,0 +1,441 @@
+        attributes = copy.deepcopy(meta)
+        del attributes[Keys.SEQUENCE_NAME]
+        del attributes[Keys.SEQUENCE_LENGTH]
+        return SequenceMetadata(name=name, length=length, index=index, attributes=attributes)
+
+    def __getitem__(self, key: str) -> str:
+        return self.attributes[key]
+
+    def __setitem__(self, key: str, value: str) -> None:
+        self.attributes[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        del self.attributes[key]
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.attributes)
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __str__(self) -> str:
+        return "@SQ\t" + "\t".join(f"{key}:{value}" for key, value in self.to_sam().items())
+
+    def __index__(self) -> int:
+        return self.index
+
+
+@dataclass(frozen=True, init=True)
+class SequenceDictionary(Mapping[Union[str, int], SequenceMetadata]):
+    """Contains an ordered collection of sequences.
+
+    Attributes:
+        infos: the ordered collection of sequence metadata
+    """
+
+    infos: List[SequenceMetadata]
+    _dict: Dict[str, SequenceMetadata] = field(init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        # Initialize a mapping from sequence name to the sequence metadata for all names
+        self_dict: Dict[str, SequenceMetadata] = {}
+        for index, info in enumerate(self.infos):
+            if info.index != index:
+                raise ValueError(
+                    "Infos must be given with index set correctly."
+                    + f"  See ${index}th with name: {info.name}"
+                )
+            for name in info.all_names:
+                if name in self_dict:
+                    raise ValueError(f"Found duplicate sequence name: {name}")
+                self_dict[name] = info
+        object.__setattr__(self, "_dict", self_dict)
+
+    def same_as(self, other: "SequenceDictionary") -> bool:
+        """Returns true if the sequences share a common reference name (including aliases), have
+        the same length, and the same MD5 if both have MD5s"""
+        if len(self) != len(other):
+            return False
+        return all(this.same_as(that) for this, that in zip(self.infos, other.infos))
+
+    def to_sam(
+        self,
+        extra_header: Optional[Dict[str, Any]] = None,
+    ) -> pysam.AlignmentHeader:
+        """Converts the sequence dictionary to a `pysam.AlignmentHeader`.
+
+        Args:
+            extra_header: a dictionary of extra values to add to the header, None otherwise.  See
+                          `:~pysam.AlignmentHeader` for more details.
+        """
+        header_dict: Dict[str, Any] = {
+            "HD": {"VN": "1.5"},
+            "SQ": [meta.to_sam() for meta in self.infos],
+        }
+        if extra_header is not None:
+            header_dict = {**header_dict, **extra_header}
+        return pysam.AlignmentHeader.from_dict(header_dict=header_dict)
```

Comment:
I think I would question here if this is the most useful thing to do.  I think it might be nicer to return just the dict generated by `[meta.to_sam() for meta in self.infos]` so it can be inserted into an existing sam header?

Reply:
Or perhaps this becomes `to_sam_file_header()`, and we also have a function that does just the above?

Reply:
Done